File: Bobble/Models/ChatSession.swift
import Foundation

enum ConversationExecutionMode: String, CaseIterable, Identifiable, Codable {
    case ask
    case bypass

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .ask:
            return "Ask"
        case .bypass:
            return "Bypass"
        }
    }

    var helpText: String {
        switch self {
        case .ask:
            return "Ask before actions that need permission."
        case .bypass:
            return "Bypass approvals for this conversation."
        }
    }

    static func defaultMode(for provider: CLIBackend) -> ConversationExecutionMode {
        switch provider {
        case .codex:
            return .bypass
        case .copilot, .claude:
            return .ask
        }
    }
}

struct ChatSession: Identifiable, Codable {
    static let defaultChatHeadSymbol = "Bobble1"
    static let availableChatHeadImageNames: [String] = (1...9).map { "Bobble\($0)" }

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case chatHeadSymbol
        case hasAssignedChatHeadSymbol
        case provider
        case selectedModel
        case conversationMode
        case messages
        case state
        case cliSessionId
        case cliSessionBackend
        case workspaceDirectory
        case createdAt
        case updatedAt
        case isArchived
    }

    let id: UUID
    var name: String
    var chatHeadSymbol: String
    var hasAssignedChatHeadSymbol: Bool
    var provider: CLIBackend
    var conversationMode: ConversationExecutionMode
    var selectedModel: ProviderModelOption
    var messages: [ChatMessage]
    var state: SessionState
    var cliSessionId: String
    var cliSessionBackend: CLIBackend?
    var workspaceDirectory: String
    var createdAt: Date
    var updatedAt: Date
    var isArchived: Bool

    enum SessionState: Codable {
        case idle
        case running
        case error(String)

        private enum CodingKeys: String, CodingKey {
            case kind
            case message
        }

        private enum Kind: String, Codable {
            case idle
            case running
            case error
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            switch try container.decode(Kind.self, forKey: .kind) {
            case .idle:
                self = .idle
            case .running:
                self = .running
            case .error:
                self = .error(try container.decode(String.self, forKey: .message))
            }
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)
            switch self {
            case .idle:
                try container.encode(Kind.idle, forKey: .kind)
            case .running:
                try container.encode(Kind.running, forKey: .kind)
            case .error(let message):
                try container.encode(Kind.error, forKey: .kind)
                try container.encode(message, forKey: .message)
            }
        }
    }

    var chatHeadImageName: String {
        guard hasAssignedChatHeadSymbol,
              let imageName = Self.sanitizedChatHeadImageName(from: chatHeadSymbol) else {
            return Self.chatHeadImageName(for: id)
        }
        return imageName
    }

    var hasUnread: Bool {
        messages.contains { $0.role == .assistant && $0.isNew }
    }

    var qualifiesForHistory: Bool {
        messages.contains { $0.role == .user }
    }

    mutating func markAssistantMessagesRead() {
        for index in messages.indices where messages[index].role == .assistant && messages[index].isNew {
            messages[index].isNew = false
        }
        touchUpdatedAt()
    }

    var attachmentsDirectory: String {
        let directoryURL = URL(fileURLWithPath: workspaceDirectory, isDirectory: true)
            .appendingPathComponent("attachments", isDirectory: true)
        try? FileManager.default.createDirectory(
            at: directoryURL,
            withIntermediateDirectories: true,
            attributes: nil
        )
        return directoryURL.path
    }

    init(
        id: UUID = UUID(),
        name: String = "New Chat",
        chatHeadSymbol: String = Self.defaultChatHeadSymbol,
        hasAssignedChatHeadSymbol: Bool = false,
        provider: CLIBackend = .codex,
        conversationMode: ConversationExecutionMode? = nil,
        selectedModel: ProviderModelOption = .automatic,
        messages: [ChatMessage] = [],
        state: SessionState = .idle,
        cliSessionId: String = UUID().uuidString,
        cliSessionBackend: CLIBackend? = nil,
        workspaceDirectory: String? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        isArchived: Bool = false
    ) {
        self.id = id
        self.name = name
        let sanitizedChatHeadSymbol = Self.sanitizedChatHeadImageName(from: chatHeadSymbol)
        self.chatHeadSymbol = sanitizedChatHeadSymbol ?? Self.defaultChatHeadSymbol
        self.hasAssignedChatHeadSymbol = hasAssignedChatHeadSymbol && sanitizedChatHeadSymbol != nil
        self.provider = provider
        self.conversationMode = conversationMode ?? ConversationExecutionMode.defaultMode(for: provider)
        self.selectedModel = selectedModel.normalized(for: provider)
        self.messages = messages
        self.state = state
        self.cliSessionId = cliSessionId
        self.cliSessionBackend = cliSessionBackend
        self.workspaceDirectory = workspaceDirectory ?? Self.createWorkspaceDirectory(for: id)
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.isArchived = isArchived
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        let id = try container.decode(UUID.self, forKey: .id)
        let name = try container.decode(String.self, forKey: .name)
        let chatHeadSymbol = try container.decodeIfPresent(String.self, forKey: .chatHeadSymbol) ?? Self.defaultChatHeadSymbol
        let hasAssignedChatHeadSymbol = try container.decodeIfPresent(Bool.self, forKey: .hasAssignedChatHeadSymbol)
            ?? (chatHeadSymbol != Self.defaultChatHeadSymbol
                && Self.sanitizedChatHeadImageName(from: chatHeadSymbol) != nil)
        let provider = try container.decode(CLIBackend.self, forKey: .provider)
        let conversationMode = try container.decodeIfPresent(ConversationExecutionMode.self, forKey: .conversationMode)
            ?? ConversationExecutionMode.defaultMode(for: provider)
        let selectedModel = try container.decodeIfPresent(ProviderModelOption.self, forKey: .selectedModel) ?? .automatic
        let messages = try container.decode([ChatMessage].self, forKey: .messages)
        let state = try container.decode(SessionState.self, forKey: .state)
        let cliSessionId = try container.decode(String.self, forKey: .cliSessionId)
        let cliSessionBackend = try container.decodeIfPresent(CLIBackend.self, forKey: .cliSessionBackend)
        let workspaceDirectory = try container.decodeIfPresent(String.self, forKey: .workspaceDirectory)
        let createdAt = try container.decodeIfPresent(Date.self, forKey: .createdAt) ?? Date()
        let updatedAt = try container.decodeIfPresent(Date.self, forKey: .updatedAt) ?? createdAt
        let isArchived = try container.decodeIfPresent(Bool.self, forKey: .isArchived) ?? false

        self.init(
            id: id,
            name: name,
            chatHeadSymbol: chatHeadSymbol,
            hasAssignedChatHeadSymbol: hasAssignedChatHeadSymbol,
            provider: provider,
            conversationMode: conversationMode,
            selectedModel: selectedModel,
            messages: messages,
            state: state,
            cliSessionId: cliSessionId,
            cliSessionBackend: cliSessionBackend,
            workspaceDirectory: workspaceDirectory,
            createdAt: createdAt,
            updatedAt: updatedAt,
            isArchived: isArchived
        )
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(chatHeadSymbol, forKey: .chatHeadSymbol)
        try container.encode(hasAssignedChatHeadSymbol, forKey: .hasAssignedChatHeadSymbol)
        try container.encode(provider, forKey: .provider)
        try container.encode(conversationMode, forKey: .conversationMode)
        try container.encode(selectedModel, forKey: .selectedModel)
        try container.encode(messages, forKey: .messages)
        try container.encode(state, forKey: .state)
        try container.encode(cliSessionId, forKey: .cliSessionId)
        try container.encodeIfPresent(cliSessionBackend, forKey: .cliSessionBackend)
        try container.encode(workspaceDirectory, forKey: .workspaceDirectory)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
        try container.encode(isArchived, forKey: .isArchived)
    }

    private static func createWorkspaceDirectory(for sessionId: UUID) -> String {
        let fm = FileManager.default

        let appSupportBase = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? fm.temporaryDirectory
        let sessionDir = appSupportBase
            .appendingPathComponent("Bobble", isDirectory: true)
            .appendingPathComponent("ChatWorkspaces", isDirectory: true)
            .appendingPathComponent(sessionId.uuidString, isDirectory: true)

        do {
            try fm.createDirectory(at: sessionDir, withIntermediateDirectories: true, attributes: nil)
            return sessionDir.path
        } catch {
            let fallback = fm.temporaryDirectory
                .appendingPathComponent("BobbleChatWorkspaces", isDirectory: true)
                .appendingPathComponent(sessionId.uuidString, isDirectory: true)
            try? fm.createDirectory(at: fallback, withIntermediateDirectories: true, attributes: nil)
            return fallback.path
        }
    }

    static func sanitizedChatHeadImageName(from rawValue: String) -> String? {
        let trimmed = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        let lowercased = trimmed.lowercased()
        guard lowercased.hasPrefix("bobble") else { return nil }
        let suffix = lowercased.dropFirst("bobble".count)
        guard let number = Int(suffix), (1...availableChatHeadImageNames.count).contains(number) else { return nil }
        return "Bobble\(number)"
    }

    static func mappedChatHeadImageName(from rawValue: String) -> String? {
        if let sanitized = sanitizedChatHeadImageName(from: rawValue) {
            return sanitized
        }

        let trimmed = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        let hash = abs(trimmed.hashValue)
        return availableChatHeadImageNames[hash % availableChatHeadImageNames.count]
    }

    private static func chatHeadImageName(for id: UUID) -> String {
        let hash = abs(id.uuidString.hashValue)
        return availableChatHeadImageNames[hash % availableChatHeadImageNames.count]
    }

    mutating func updateChatHeadSymbol(from rawValue: String?) {
        guard let rawValue,
              let imageName = Self.mappedChatHeadImageName(from: rawValue) else {
            return
        }

        chatHeadSymbol = imageName
        hasAssignedChatHeadSymbol = true
        touchUpdatedAt()
    }

    mutating func touchUpdatedAt() {
        updatedAt = Date()
    }

    func normalizedForRestore() -> ChatSession {
        var restored = self
        if case .running = restored.state {
            restored.state = .idle
        }
        for index in restored.messages.indices {
            if restored.messages[index].isStreaming {
                restored.messages[index].isStreaming = false
            }
            if !restored.messages[index].interruptionActions.isEmpty {
                restored.messages[index].interruptionActions = []
            }
        }
        return restored
    }
}

File: BobbleWin/BobbleWin/Models/ChatSession.cs
using System.Text.Json.Serialization;

namespace BobbleWin.Models;

public enum ConversationExecutionMode
{
    Ask,
    Bypass,
}

public static class ConversationExecutionModeExtensions
{
    public static string DisplayName(this ConversationExecutionMode mode) => mode switch
    {
        ConversationExecutionMode.Ask => "Ask",
        ConversationExecutionMode.Bypass => "Bypass",
        _ => "Ask",
    };

    public static string HelpText(this ConversationExecutionMode mode) => mode switch
    {
        ConversationExecutionMode.Ask => "Ask before actions that need permission.",
        ConversationExecutionMode.Bypass => "Bypass approvals for this conversation.",
        _ => string.Empty,
    };

    public static ConversationExecutionMode DefaultMode(CLIBackend backend) => backend switch
    {
        CLIBackend.Codex => ConversationExecutionMode.Bypass,
        CLIBackend.Copilot => ConversationExecutionMode.Ask,
        CLIBackend.Claude => ConversationExecutionMode.Ask,
        _ => ConversationExecutionMode.Ask,
    };
}

public enum SessionStateKind
{
    Idle,
    Running,
    Error,
}

public sealed class SessionState
{
    public SessionStateKind Kind { get; set; } = SessionStateKind.Idle;
    public string? Message { get; set; }

    public static SessionState Idle() => new() { Kind = SessionStateKind.Idle };
    public static SessionState Running() => new() { Kind = SessionStateKind.Running };
    public static SessionState Error(string message) => new() { Kind = SessionStateKind.Error, Message = message };
}

public sealed class ChatSession
{
    public const string DefaultChatHeadSymbol = "Bobble1";
    public static readonly IReadOnlyList<string> AvailableChatHeadImageNames =
        Enumerable.Range(1, 9).Select(value => $"Bobble{value}").ToList();

    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "New Chat";
    public string ChatHeadSymbol { get; set; } = DefaultChatHeadSymbol;
    public bool HasAssignedChatHeadSymbol { get; set; }
    public CLIBackend Provider { get; set; } = CLIBackend.Codex;
    public ConversationExecutionMode ConversationMode { get; set; } = ConversationExecutionMode.Bypass;
    public ProviderModelOption SelectedModel { get; set; } = ProviderModelOption.Automatic;
    public List<ChatMessage> Messages { get; set; } = [];
    public SessionState State { get; set; } = SessionState.Idle();
    public string CliSessionId { get; set; } = Guid.NewGuid().ToString();
    public CLIBackend? CliSessionBackend { get; set; }
    public string WorkspaceDirectory { get; set; } = string.Empty;
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.Now;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.Now;
    public bool IsArchived { get; set; }

    [JsonIgnore]
    public string ChatHeadImageName
    {
        get
        {
            if (HasAssignedChatHeadSymbol)
            {
                var sanitized = SanitizedChatHeadImageName(ChatHeadSymbol);
                if (!string.IsNullOrWhiteSpace(sanitized))
                {
                    return sanitized;
                }
            }

            var index = Math.Abs(Id.GetHashCode()) % AvailableChatHeadImageNames.Count;
            return AvailableChatHeadImageNames[index];
        }
    }

    [JsonIgnore]
    public bool HasUnread => Messages.Any(message => message.Role == ChatMessageRole.Assistant && message.IsNew);

    [JsonIgnore]
    public bool QualifiesForHistory => Messages.Any(message => message.Role == ChatMessageRole.User);

    public string AttachmentsDirectory()
    {
        var path = Path.Combine(WorkspaceDirectory, "attachments");
        Directory.CreateDirectory(path);
        return path;
    }

    public static ChatSession Create(CLIBackend provider, string? workspaceDirectory = null)
    {
        var sessionId = Guid.NewGuid();
        return new ChatSession
        {
            Id = sessionId,
            Name = "New Chat",
            ChatHeadSymbol = DefaultChatHeadSymbol,
            HasAssignedChatHeadSymbol = false,
            Provider = provider,
            ConversationMode = ConversationExecutionModeExtensions.DefaultMode(provider),
            SelectedModel = ProviderModelOption.Automatic,
            Messages = [],
            State = SessionState.Idle(),
            CliSessionId = Guid.NewGuid().ToString(),
            CliSessionBackend = null,
            WorkspaceDirectory = workspaceDirectory ?? CreateWorkspaceDirectory(sessionId),
            CreatedAt = DateTimeOffset.Now,
            UpdatedAt = DateTimeOffset.Now,
            IsArchived = false,
        };
    }

    public void EnsureDefaults()
    {
        if (string.IsNullOrWhiteSpace(WorkspaceDirectory))
        {
            WorkspaceDirectory = CreateWorkspaceDirectory(Id);
        }

        ConversationMode = ConversationMode;
        SelectedModel = SelectedModel.Normalized(Provider);
    }

    public void MarkAssistantMessagesRead()
    {
        foreach (var message in Messages.Where(message => message.Role == ChatMessageRole.Assistant && message.IsNew))
        {
            message.IsNew = false;
        }

        TouchUpdatedAt();
    }

    public void UpdateChatHeadSymbol(string? rawValue)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return;
        }

        var mapped = MappedChatHeadImageName(rawValue);
        if (string.IsNullOrWhiteSpace(mapped))
        {
            return;
        }

        ChatHeadSymbol = mapped;
        HasAssignedChatHeadSymbol = true;
        TouchUpdatedAt();
    }

    public void TouchUpdatedAt()
    {
        UpdatedAt = DateTimeOffset.Now;
    }

    public ChatSession NormalizedForRestore()
    {
        var copy = Clone();
        if (copy.State.Kind == SessionStateKind.Running)
        {
            copy.State = SessionState.Idle();
        }

        foreach (var message in copy.Messages)
        {
            message.IsStreaming = false;
            if (message.InterruptionActions.Count > 0)
            {
                message.InterruptionActions = [];
            }
        }

        return copy;
    }

    public ChatSession Clone()
    {
        return new ChatSession
        {
            Id = Id,
            Name = Name,
            ChatHeadSymbol = ChatHeadSymbol,
            HasAssignedChatHeadSymbol = HasAssignedChatHeadSymbol,
            Provider = Provider,
            ConversationMode = ConversationMode,
            SelectedModel = SelectedModel,
            Messages = Messages.Select(CloneMessage).ToList(),
            State = new SessionState { Kind = State.Kind, Message = State.Message },
            CliSessionId = CliSessionId,
            CliSessionBackend = CliSessionBackend,
            WorkspaceDirectory = WorkspaceDirectory,
            CreatedAt = CreatedAt,
            UpdatedAt = UpdatedAt,
            IsArchived = IsArchived,
        };
    }

    private static ChatMessage CloneMessage(ChatMessage message)
    {
        return new ChatMessage
        {
            Id = message.Id,
            Role = message.Role,
            Content = message.Content,
            Attachments = message.Attachments.Select(attachment => new ChatAttachment
            {
                Id = attachment.Id,
                Kind = attachment.Kind,
                FileName = attachment.FileName,
                FilePath = attachment.FilePath,
                RelativePath = attachment.RelativePath,
            }).ToList(),
            InterruptionTitle = message.InterruptionTitle,
            InterruptionDetails = message.InterruptionDetails,
            InterruptionActions = message.InterruptionActions.Select(action => new InterruptionAction
            {
                Id = action.Id,
                Title = action.Title,
                Role = action.Role,
                Payload = action.Payload,
            }).ToList(),
            Timestamp = message.Timestamp,
            IsStreaming = message.IsStreaming,
            IsNew = message.IsNew,
            Kind = message.Kind,
        };
    }

    public static string? SanitizedChatHeadImageName(string rawValue)
    {
        var trimmed = rawValue.Trim();
        if (trimmed.Length == 0)
        {
            return null;
        }

        var lower = trimmed.ToLowerInvariant();
        if (!lower.StartsWith("bobble", StringComparison.Ordinal))
        {
            return null;
        }

        var suffix = lower["bobble".Length..];
        if (!int.TryParse(suffix, out var number))
        {
            return null;
        }

        if (number < 1 || number > AvailableChatHeadImageNames.Count)
        {
            return null;
        }

        return $"Bobble{number}";
    }

    public static string? MappedChatHeadImageName(string rawValue)
    {
        var sanitized = SanitizedChatHeadImageName(rawValue);
        if (!string.IsNullOrWhiteSpace(sanitized))
        {
            return sanitized;
        }

        var trimmed = rawValue.Trim();
        if (trimmed.Length == 0)
        {
            return null;
        }

        var index = Math.Abs(trimmed.GetHashCode()) % AvailableChatHeadImageNames.Count;
        return AvailableChatHeadImageNames[index];
    }

    private static string CreateWorkspaceDirectory(Guid sessionId)
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        var preferred = Path.Combine(appData, "BobbleWin", "ChatWorkspaces", sessionId.ToString());
        try
        {
            Directory.CreateDirectory(preferred);
            return preferred;
        }
        catch
        {
            var fallback = Path.Combine(Path.GetTempPath(), "BobbleWinChatWorkspaces", sessionId.ToString());
            Directory.CreateDirectory(fallback);
            return fallback;
        }
    }
}
